assets/hdfs_append/index.js

```diff
@@ -0,0 +1,161 @@
+                    let newFilename = '';
+                    if (!appendErrors[filename]) {
+                        newFilename = `${filename}.0`;
+                        appendErrors[filename] = newFilename;
+                    } else {
+                        // Get the last attempted file and increment the number
+                        const incNum = appendErrors[filename].split('.').reverse()[0] * 1 + 1;
+                        newFilename = `${filename}.${incNum}`;
+                        // Set the new target for the next slice attempt
+                        appendErrors[filename] = newFilename;
+                    }
+                    sliceError = `Error sending data to file '${filename}' due to HDFS append `
+                        + `error. Changing destination to '${newFilename}'. Error: ${errMsg}`;
+                } else {
+                    sliceError = `Error sending data to file: ${filename}, error: ${errMsg}`;
+                }
+                if (opConfig.log_data_on_error === true) {
+                    sliceError = `${sliceError} Data: ${JSON.stringify(chunks)}`;
+                }
+                return Promise.reject(sliceError);
+            });
+    }
+
+    return (data) => {
+        // Start by mapping data chunks to their respective files
+        const map = {};
+        data.forEach((record) => {
+            // This skips any records that have non-existant data payloads to avoid empty appends
+            if (record.data.length > 0) {
+                let file = record.filename;
+                // If the file has already had an error, update the filename for the next write
+                // attempt
+                if (appendErrors[file]) {
+                    const attemptNum = appendErrors[file].split('.').reverse()[0] * 1;
+                    // This stops the worker from creating too many new files
+                    if (attemptNum > opConfig.max_writes) {
+                        throw new Error(
+                            `${file} has exceeded the maximum number of write attempts!`
+                        );
+                    }
+                    file = appendErrors[file];
+                }
+                if (!map[file]) map[file] = [];
+                map[file].push(record.data);
+            }
+        });
+
+        function sendFiles() {
+            const stores = [];
+            _.forOwn(map, (chunks, key) => {
+                stores.push(prepareFile(key, chunks, jobConfig.logger));
+            });
+
+            // We can process all individual files in parallel.
+            return Promise.all(stores)
+                .catch((err) => {
+                    const errMsg = err.stack ? err.stack : err;
+                    jobConfig.logger.error(`Error while sending to hdfs, error: ${errMsg}`);
+                    return Promise.reject(err);
+                });
+        }
+
+        return sendFiles();
+    };
+}
+
+function getClient(context, config, type) {
```

Comment:
I believe you can get this function directly from Teraslice now so this copy could be removed. Check with Jared on how to do it.